File: skills/code-architect/references/code-examples/php/FactoryMethod.php
<?php
// Factory Method — subclasses decide which product class to instantiate via a factory method.
declare(strict_types=1);

interface Transport
{
    public function describeRoute(): string;
}

final class LandTransport implements Transport
{
    public function describeRoute(): string
    {
        return 'delivering over roads';
    }
}

final class WaterTransport implements Transport
{
    public function describeRoute(): string
    {
        return 'delivering across the sea';
    }
}

abstract class Dispatcher
{
    // The factory method: each subclass returns its own product.
    abstract protected function createTransport(): Transport;

    public function planDelivery(): string
    {
        $transport = $this->createTransport();
        return 'Dispatcher is ' . $transport->describeRoute();
    }
}

final class RoadDispatcher extends Dispatcher
{
    protected function createTransport(): Transport
    {
        return new LandTransport();
    }
}

final class SeaDispatcher extends Dispatcher
{
    protected function createTransport(): Transport
    {
        return new WaterTransport();
    }
}

function clientCode(Dispatcher $dispatcher): void
{
    echo $dispatcher->planDelivery() . PHP_EOL;
}

// Demo
clientCode(new RoadDispatcher());
clientCode(new SeaDispatcher());

File: skills/code-architect/references/code-examples/csharp/FactoryMethod.cs
// Factory Method — subclasses decide which product class to instantiate via a factory method.
using System;

namespace Patterns.FactoryMethod
{
    interface IProduct
    {
        string Describe();
    }

    sealed class ConcreteProductA : IProduct
    {
        public string Describe() => "Product A";
    }

    sealed class ConcreteProductB : IProduct
    {
        public string Describe() => "Product B";
    }

    abstract class Creator
    {
        // The factory method: subclasses supply the concrete product.
        protected abstract IProduct CreateProduct();

        // Higher-level logic relying on the product, agnostic of its concrete type.
        public string Operate()
        {
            IProduct product = CreateProduct();
            return $"Creator built [{product.Describe()}]";
        }
    }

    sealed class CreatorA : Creator
    {
        protected override IProduct CreateProduct() => new ConcreteProductA();
    }

    sealed class CreatorB : Creator
    {
        protected override IProduct CreateProduct() => new ConcreteProductB();
    }

    static class Demo
    {
        public static void Run()
        {
            Creator[] creators = { new CreatorA(), new CreatorB() };
            foreach (Creator creator in creators)
                Console.WriteLine(creator.Operate());
        }

        public static void Main() => Run();
    }
}
